zproject/backends.py

```diff
@@ -380,7 +380,9 @@ def sync_custom_profile_fields_from_ldap(self, user_profile: UserProfile,
 
     def get_or_build_user(self, username: str,
                           ldap_user: _LDAPUser) -> Tuple[UserProfile, bool]:
-        (user, built) = super().get_or_build_user(username, ldap_user)
+        assert self._realm is not None
```

Comment:
I think this will break the `POPULATE_PROFILE_VIA_LDAP` setting's use case on account creation, which is a way you can have Zulip manage auth/passwords but have LDAP manage users' names.  (Note that here we're using the backend instance initialized in Django setup, which won't have an `_realm` yet).

```
        elif settings.POPULATE_PROFILE_VIA_LDAP:                                                                                        
            for backend in get_backends():                                                                                              
                if isinstance(backend, LDAPBackend):                                                                                    
                    try:                                                                                                                
                        ldap_username = backend.django_to_ldap_username(email)                                                          
                    except ZulipLDAPExceptionOutsideDomain:                                                                             
                        logging.warning("New account email %s could not be found in LDAP" % (email,))                                   
                        form = RegistrationForm(realm_creation=realm_creation)                                                          
                        break                                                                                                           
                                                                                                                                        
                    ldap_user = _LDAPUser(backend, ldap_username)                                                                       
                                                                                                                                        
                    try:                                                                                                                
                        ldap_full_name, _ = backend.get_mapped_name(ldap_user)                                                          
                        request.session['authenticated_full_name'] = ldap_full_name                                                     
                        name_validated = True                                                                                           
                        # We don't use initial= here, because if the form is                                                            
                        # complete (that is, no additional fields need to be                                                            
                        # filled out by the user) we want the form to validate,                                                         
                        # so they can be directly registered without having to                                                          
                        # go through this interstitial.                                                                                 
                        form = RegistrationForm({'full_name': ldap_full_name},                                                          
                                                realm_creation=realm_creation)                                                          
                        require_ldap_password = True                                                                                    
                        break                                                                                                           
                    except TypeError:                                                                                                   
                        # Let the user fill out a name and/or try another backend                                                       
                        form = RegistrationForm(realm_creation=realm_creation)                                                          
```